server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java

```diff
@@ -45,4 +58,117 @@ public void testZen1NodesJoiningZen2Cluster() {
         internalCluster().startNodes(randomIntBetween(1, 3), ZEN1_SETTINGS);
         createIndex("test");
     }
+
+    public void testMixedClusterFormation() throws Exception {
+        final int zen1NodeCount = randomIntBetween(1, 3);
+        final int zen2NodeCount = randomIntBetween(1, 3);
+        logger.info("starting cluster of [{}] Zen1 nodes and [{}] Zen2 nodes", zen1NodeCount, zen2NodeCount);
+        final List<String> nodes = internalCluster().startNodes(IntStream.range(0, zen1NodeCount + zen2NodeCount)
+            .mapToObj(i -> i < zen1NodeCount ? ZEN1_SETTINGS : ZEN2_SETTINGS).toArray(Settings[]::new));
+
+        for (final String node : nodes) {
+            if (zen1NodeCount == 1 && node.equals(nodes.get(0))) {
+                // Do not restart the only Zen1 node, because it cannot recover state from the Zen2 nodes. This is actually kinda bad,
+                // if the Zen1 node restarts quickly enough then the Zen2 nodes will not bootstrap themselves and will try and join the
+                // Zen1 node but they won't be able to because its cluster state version is 1.
```

Comment:
Note that the same issue exists if you have migrated some non-master-eligible nodes to Zen2, and then do a full restart of the Zen1 masters.